<a href="https://colab.research.google.com/github/jasimrashid/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

In [0]:
import pandas as pd
from scipy import stats
import numpy as np

In [2]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-04-13 19:10:53--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.03s   

2020-04-13 19:10:54 (634 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [4]:
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df=pd.read_csv('house-votes-84.data', names=column_headers)

print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [8]:
df=df.replace({'y':1,'n':0,'?':np.NaN})
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
dem = df[df['party']=='democrat']
rep = df[df['party']=='republican']

In [0]:
dem.head()

**Answers**

1. Find an issue that democrats support more than republicans with p < 0.01: adoption-of-the-budget-resolution
2. Find an issue that republicans support more than democrats with p < 0.01: religious-groups-in-schools
3. Find an issue where the difference between republicans and democrats has p > 0.1: Water-project-cost-sharing

**Issue #1: Water Project (issue where the difference between republicans and democrats has p > 0.1)**

Step 1: Null: Sample means are equal

Step 2: Alternative: Sample means are different

Step 3: Confidence Level: 95% *italicized text*

In [13]:
# Step 4: Calculate p-value 
stats.ttest_ind(dem['water-project'], rep['water-project'], nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

**Step 5: Conclusion**: Based on t-statistic of -0.089 and p-value of 0.929, I fail to reject the null hypothesis that democrats' support for water project bill is the same as republicans

In [0]:
# print(dem['water-project'].mean() - rep['water-project'].mean())

**Issue #2: Adoption of Budget Resolution (democrats support more than republicans with p < 0.01)**

Step 1: Null: Sample means are equal

Step 2: Alternative: Sample means are different

Step 3: Confidence Level: 95%

In [18]:
# Step 4: Calculate p-value 
stats.ttest_ind(dem['budget'], rep['budget'], nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

**Conclusion**: Based on t-statistic of 23.213 and p-value of 2.070 x exp(-77), I reject the null hypothesis that democrats' support for water project bill is different. A positive t-value indicates that **democrats are more likely than republicans to support the budget resolution.**

**Issue #3: Religious Groups in Schools**

Step 1: Null: Sample means are equal

Step 2: Alternative: Sample means are different

Step 3: Confidence Level: 95%

In [19]:
# Step 4: Calculate p-value 
stats.ttest_ind(dem['religious-groups'], rep['religious-groups'], nan_policy='omit')

Ttest_indResult(statistic=-9.737575825219457, pvalue=2.3936722520597287e-20)

**Conclusion**: Based on t-statistic of -9.738 and p-value of 2.394 x exp(-20), I reject the null hypothesis that democrats' support for water project bill is different. A negative t-value indicates that **republicans are more likely than democrats to support the budget resolution.**


## Stretch Goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then check your Answers using Scipy!

In [0]:
# Define functions for 1-sample t-test and 2-sample t-test

# 1-sample t-test
def one_sample_t_test(x, pop_mean):
    
    x_mean = x.dropna().mean()
    std = x.dropna().std()
    n = len(x.dropna())
    t = (x_mean-pop_mean)/(std/((n)**0.5))
    print(t)


def two_sample_t_test(x, y):
    
    x_mean = x.dropna().mean()
    y_mean = y.dropna().mean()
    n1 = len(x.dropna())
    n2 = len(y.dropna())
    s = ((((x.dropna()-x_mean)**2).sum() + ((y.dropna()-y_mean)**2).sum())/(n1+n2-2))**0.5
    t = (x_mean-y_mean)/(((s**2)*((1/n1)+(1/n2)))**0.5)
    print(t)


In [126]:
# Execute 1-sample test

# **Set parameters below**
column_name = 'anti-satellite-ban'
pop_mean = .8
dem_or_rep = 'rep'


if column_name in dem.columns.to_list():
  if dem_or_rep == 'dem':
    one_sample_t_test(dem[column_name],pop_mean)
  else:
    one_sample_t_test(rep[column_name],pop_mean)
else:
  print("Invalid column name")

-16.598019728624834


In [0]:
# t = -16.59 translates to a p-value lower than 0.0005 based on t-stat table

In [137]:
# Execute 2-sample test 

# **Set parameters below**
column_name = 'anti-satellite-ban'


if column_name in dem.columns.to_list():
    two_sample_t_test(dem[column_name],rep[column_name])
else:
  print("Invalid column name")


Ttest_indResult(statistic=12.526187929077842, pvalue=8.521033017443867e-31)
12.526187929077844


In [0]:
# t = 12.53 translates to a p-value lower than 0.0005 based on t-stat table